File: ios/BleNative/Handlers/RequestHandler.swift
import Foundation

struct RequestTimeout {
    let timeout: Int
    let onTimeout: (Request) -> Void 
}

class RequestHandler {
    private(set) var requests: [Request] = []
    private var timeoutsToRequest: [Request: DispatchWorkItem] = [:]
    
    func addRequest(_ request: Request, timeout: RequestTimeout?) {
        requests.append(request)
        if let timeout = timeout {
            addTimeout(timeout.timeout, forRequest: request, onTimeout: timeout.onTimeout)
        }
    }
    
    private func removeRequest(atIndex index: Int) -> Request? {
        let request = requests.remove(at: index)
        if let task = timeoutsToRequest.removeValue(forKey: request) {
            task.cancel()
        }
        return request
    }
    
    @discardableResult
    func removeRequest(promiseId: String) -> Request? {
        guard let index = requests.index(where: { $0.promiseId == promiseId }) else {
            return nil
        }
        return removeRequest(atIndex: index)
    }
    
    @discardableResult
    func removeRequest(relatedIdentifier: Int32, type: RequestType) -> Request? {
        guard let index = requests.index(where: { $0.relatedIdentifier == relatedIdentifier && $0.type == type }) else {
            return nil
        }
        return removeRequest(atIndex: index)
    }
    
    @discardableResult
    func removeRequest(_ request: Request) -> Request? {
        guard let index = requests.index(of: request) else { return nil }
        return removeRequest(atIndex: index)
    }
    
    func findRequest(relatedIdentifier: Int32, type: RequestType) -> Request? {
        return requests.first(where: { $0.relatedIdentifier == relatedIdentifier && $0.type == type })
    }
    
    private func addTimeout(_ timeout: Int, forRequest request: Request, onTimeout: @escaping (Request) -> Void) {
        let task = DispatchWorkItem { [weak self] in
            self?.removeRequest(request)
            onTimeout(request)            
        }
        timeoutsToRequest[request] = task
        
        let deadlineTime = DispatchTime.now() + .milliseconds(timeout)
        DispatchQueue.main.asyncAfter(deadline: deadlineTime, execute: task)
    }
}

File: android/src/main/java/com/polidea/blenative/handlers/RequestHandler.kt
package com.polidea.blenative.handlers

import com.polidea.blenative.models.Request
import com.polidea.blenative.models.RequestType
import com.polidea.blenative.task.TaskDispatcher

data class RequestTimeout(val timeout: Int, val onTimeout: (Request) -> Void)

class RequestHandler(private val taskDispatcher: TaskDispatcher) {

    private val requests = mutableListOf<Request>()

    private val timeoutRunnablesToRequest = mutableMapOf<Request, Runnable>()

    fun addRequest(request: Request, timeout: RequestTimeout?) {
        requests.add(request)
        timeout?.let {
            addTimeout(timeout.timeout, request, timeout.onTimeout)
        }
    }

    fun removeRequest(promiseId: String): Request? {
        val index = requests.indexOfFirst { it.promiseId == promiseId }
        return if (index == -1) {
            null
        } else {
            removeRequest(index)
        }
    }

    fun removeRequest(relatedIdentifier: Int, type: RequestType): Request? {
        val index = requests.indexOfFirst { it.relatedIdentifier == relatedIdentifier && it.type == type }
        if (index == -1) {
            return null
        }
        return removeRequest(index)
    }

    fun removeRequest(request: Request): Request? {
        val index = requests.indexOf(request)
        if (index == -1) {
            return null
        }
        return removeRequest(index)
    }

    private fun removeRequest(index: Int): Request? {
        val request = requests.removeAt(index)
        val timeoutRunnable = timeoutRunnablesToRequest.remove(request)
        timeoutRunnable?.let {
            taskDispatcher.removeCallbacks(it)
        }
        return request

    }

    fun findRequest(relatedIdentifier: Int, type: RequestType): Request? {
        return requests.first { it.relatedIdentifier == relatedIdentifier && it.type == type }
    }

    private fun addTimeout(timeout: Int, request: Request, onTimeout: (Request) -> Void) {
        val runnable = Runnable {
            removeRequest(request)
            onTimeout(request)
        }
        timeoutRunnablesToRequest[request] = runnable

        taskDispatcher.postDelayed(runnable, (timeout * 1000).toLong())
    }
}